### ISSUE TITLE
Curl configured with cmake doesn't work with libidn2 even though it has enabled IDN support

### ISSUE THREAD

**Issue by dmilith** [curl/curl#9022]
### I did this
Configured Curl using cmake v3.19.3

### I expected the following
The configuration phase using autotools works properly and detects IDN2 library.

When I configure with cmake, I get

// Use libidn2 for IDN support
USE_LIBIDN2:BOOL=ON

info from configuration part, but after build the curl -V doesn't show "IDN" feature… also doesn't work with IDN domains (tested HTTP://żółw.pl/ in my case)

### curl/libcurl version
7.83.1

[curl -V output]

cmake configured curl:

curl 7.83.1 (FreeBSD) libcurl/7.83.1 LibreSSL/3.4.3 zlib/1.2.12 brotli/1.0.9 zstd/1.5.2 nghttp2/1.47.0
Release-Date: 2022-05-11
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM SSL UnixSockets zstd

and the same output from the autotools-configured curl:

curl 7.78.0 (aarch64-unknown-freebsd13.0) libcurl/7.78.0 LibreSSL/3.2.5 zlib/1.2.12 libidn2/2.3.2 libssh2/1.9.0 nghttp2/1.44.0
Release-Date: 2021-07-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz NTLM NTLM_WB SSL UnixSockets

### operating system

FreeBSD vks5 13.1-RELEASE FreeBSD 13.1-RELEASE releng/13.1-n250148-fc952ac2212 GENERIC arm64

**Comment by dmilith** [comment]
(please note that the build configurations are not exactly the same, I added some additional features since I switched to the CMake build system).

**Comment by bagder** [comment]
I just tried this on my debian Linux on git master:
~~~
$ cd build
$ cmake -DUSE_LIBIDN2=ON ..
$ make -sj7
$ ./src/curl -V
curl 7.84.0-DEV (Linux) libcurl/7.84.0-DEV OpenSSL/3.0.3 zlib/1.2.11 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.10.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IDN IPv6 Largefile libz NTLM PSL SSL UnixSockets
~~~

I take it this is not what happens for you? Does it detect the idn2.h header file fine for you?

**Comment by dmilith** [comment]
@bagder Hey, thanks for super fast response :) Well, the full configuration block is here: https://gist.github.com/dmilith/9c97880ce96ea6eca7f0a3948728641c

and it contains the 

```
-- Looking for 21 include files stdio.h, ..., idn2.h
-- Looking for 21 include files stdio.h, ..., idn2.h - found
```

and

```
// Use libidn2 for IDN support
USE_LIBIDN2:BOOL=ON
```

**Comment by vszakats** [comment]
I didn't mean to fix this specific issue with that patch. It's worth a try to see if it changed anything for this though.

libidn2 still seems to (sometimes?) requires a fully manual setup, along the lines of:
```shell
'-DUSE_LIBIDN2=ON'
'-DCMAKE_C_FLAGS=-I/path/to/libidn2/include'
'-DCMAKE_EXE_LINKER_FLAGS=-L/path/to/libidn2/lib -lidn2'
'-DCMAKE_SHARED_LINKER_FLAGS=-L/path/to/libidn2/lib -lidn2'
```

`[LIB]IDN2_LIBRARY`, `[LIB]IDN2_INCLUDE_DIR` settings did not work for me. Detection uses `check_library_exists_concat()`, while most other similar dependencies seem to rely on `find_package()`.

**Comment by dmilith** [comment]
I've checked your suggestion, and extended my cmake configuration params with: `-DCMAKE_C_FLAGS=-I${PREFIX}/include -DCMAKE_EXE_LINKER_FLAGS='-L${PREFIX}/lib -lidn2' -DCMAKE_SHARED_LINKER_FLAGS='-L${PREFIX}/lib -lidn2'`, but the result is the same - no IDN feature in the product binary.

**Comment by vszakats** [comment]
@dmilith Did you apply the patch also? It is possible my manual configuration doesn't work without it.

**Comment by dmilith** [comment]
> @dmilith Did you apply the patch also? It is possible my manual configuration doesn't work without it.

Silly me! You're right, after applying the patch it worked as expected!

```
curl 7.83.1 (FreeBSD) libcurl/7.83.1 LibreSSL/3.4.3 zlib/1.2.12 brotli/1.0.9 zstd/1.5.2 libidn2/2.3.2 nghttp2/1.47.0
Release-Date: 2022-05-11
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz NTLM SSL UnixSockets zstd
```

**Comment by Keitagit-kun** [comment]
I have pretty same problem with static idn2 library (not found):
from cmake-gui  (cmake 3.10.1)

wich patch:
...
Looking for idn2_lookup_ul in idn2;wldap32;winmm;ws2_32;C:/Temp/curl 7.83.1-OpenSSL3/deps/lib/libcares.a;OpenSSL::SSL;OpenSSL::Crypto;C:/Temp/curl 7.83.1-OpenSSL3/deps/lib/libnghttp2.a
Looking for idn2_lookup_ul in idn2;wldap32;winmm;ws2_32;C:/Temp/curl 7.83.1-OpenSSL3/deps/lib/libcares.a;OpenSSL::SSL;OpenSSL::Crypto;C:/Temp/curl 7.83.1-OpenSSL3/deps/lib/libnghttp2.a - not found
...
Looking for 13 include files C:/Temp/curl 7.83.1-OpenSSL3/deps/include/libssh2.h, ..., idn2.h
Looking for 13 include files C:/Temp/curl 7.83.1-OpenSSL3/deps/include/libssh2.h, ..., idn2.h - found
...
--------------------------------
wichout patch:
...
Looking for 13 include files C:/Temp/curl 7.83.1-OpenSSL3/deps/include/libssh2.h, ..., idn2.h
Looking for 13 include files C:/Temp/curl 7.83.1-OpenSSL3/deps/include/libssh2.h, ..., idn2.h - found
...
--------------------------------
configuration rezult in both cases is the same - there a no IDN entry:

Enabled features: SSL IPv6 unixsockets libz brotli zstd AsynchDNS Largefile alt-svc HSTS NTLM HTTP2 HTTPS-proxy unicode
Enabled protocols: DICT FILE FTP FTPS GOPHER GOPHERS HTTP HTTPS IMAP IMAPS LDAP MQTT POP3 POP3S RTSP SCP SFTP SMB SMBS SMTP SMTPS TELNET TFTP
Enabled SSL backends: OpenSSL
Configuring done

it seems static idn2 lirary still not found in .../deps/lib/libidn2.a

**Comment by vszakats** [comment]
@Keitagit-kun Have you tried renaming `C:/Temp/curl 7.83.1-OpenSSL3/` to not have spaces?

My log snippet, with the patch, and the manual config mentioned above:
```
[...]
+ cmake . -DCMAKE_SYSTEM_NAME=Windows -DCMAKE_BUILD_TYPE=Release -DCMAKE_DISABLE_PRECOMPILE_HEADERS=OFF -DCMAKE_AR=/usr/local/opt/mingw-w64/toolchain-x86_64/bin/x86_64-w64-mingw32-ar -DCMAKE_RC_COMPILER=x86_64-w64-mingw32-windres -DCMAKE_INSTALL_MESSAGE=NEVER -DCMAKE_INSTALL_PREFIX=/usr/local -DCURL_TARGET_WINDOWS_VERSION=0x0600 -DCURL_OS_SUFFIX=-x64 -DCMAKE_SHARED_LIBRARY_SUFFIX_C=-x64.dll -DUSE_ZLIB=ON -DZLIB_LIBRARY=/my-curl/curl/../zlib/pkg/usr/local/lib/libz.a -DZLIB_INCLUDE_DIR=/my-curl/curl/../zlib/pkg/usr/local/include -DCURL_BROTLI=ON -DBROTLIDEC_LIBRARY=/my-curl/curl/../brotli/pkg/usr/local/lib/libbrotlidec.a -DBROTLICOMMON_LIBRARY=/my-curl/curl/../brotli/pkg/usr/local/lib/libbrotlicommon.a -DCURL_USE_OPENSSL=ON -DOPENSSL_ROOT_DIR=/my-curl/curl/../openssl-quic/pkg/usr/local -DOPENSSL_INCLUDE_DIR=/my-curl/curl/../openssl-quic/pkg/usr/local/include -DCURL_DISABLE_OPENSSL_AUTO_LOAD_CONFIG=ON -DCURL_USE_SCHANNEL=ON -DCURL_USE_LIBSSH2=ON -DLIBSSH2_LIBRARY=/my-curl/curl/../libssh2/pkg/usr/local/lib/libssh2.a -DLIBSSH2_INCLUDE_DIR=/my-curl/curl/../libssh2/pkg/usr/local/include -DUSE_NGHTTP2=ON -DNGHTTP2_LIBRARY=/my-curl/curl/../nghttp2/pkg/usr/local/lib/libnghttp2.a -DNGHTTP2_INCLUDE_DIR=/my-curl/curl/../nghttp2/pkg/usr/local/include -DUSE_NGHTTP3=ON -DNGHTTP3_LIBRARY=/my-curl/curl/../nghttp3/pkg/usr/local/lib/libnghttp3.a -DNGHTTP3_INCLUDE_DIR=/my-curl/curl/../nghttp3/pkg/usr/local/include -DUSE_NGTCP2=ON -DNGTCP2_LIBRARY=/my-curl/curl/../ngtcp2/pkg/usr/local/lib/libngtcp2.a -DNGTCP2_INCLUDE_DIR=/my-curl/curl/../ngtcp2/pkg/usr/local/include -DCMAKE_LIBRARY_PATH=/my-curl/curl/../ngtcp2/pkg/usr/local/lib -DUSE_LIBIDN2=ON -DENABLE_MANUAL=ON -DCURL_CA_PATH=none -DCURL_CA_BUNDLE=none -DBUILD_SHARED_LIBS=ON -DBUILD_CURL_EXE=OFF -DENABLE_THREADED_RESOLVER=ON -DBUILD_TESTING=OFF -DCMAKE_SYSROOT=/usr/local/opt/mingw-w64/toolchain-x86_64 -DCMAKE_LIBRARY_ARCHITECTURE=x86_64-w64-mingw32 -DCMAKE_C_COMPILER_TARGET=x86_64-w64-mingw32 -DCMAKE_C_COMPILER=clang '-DCMAKE_C_FLAGS=-fno-ident -W -Wall -DHAVE_STRCASECMP -DHAVE_ATOMIC -DHAVE_SIGNAL -DHAVE_SOCKADDR_IN6_SIN6_SCOPE_ID -DHAVE_STRTOK_R -DUSE_HEADERS_API -DHAVE_FTRUNCATE -DHAVE_GETADDRINFO_THREADSAFE -DHAVE_UNISTD_H -DHAVE_STRUCT_POLLFD -D_UCRT -DHAVE_LDAP_SSL -DHAVE_OPENSSL_SRP -DUSE_TLS_SRP -DHAS_ALPN -DNGHTTP2_STATICLIB -DNGHTTP3_STATICLIB -DNGTCP2_STATICLIB -DUSE_GSASL -I/my-curl/curl/../libgsasl/pkg/usr/local/include -I/my-curl/curl/../libidn2/pkg/usr/local/include -DUSE_MANUAL=1' '-DCMAKE_EXE_LINKER_FLAGS=-static-libgcc -Wl,--nxcompat -Wl,--dynamicbase -Wl,--high-entropy-va -fuse-ld=lld -s -lucrt -Wl,--start-group -lwldap32 -lbcrypt -lbcrypt -lws2_32 -L/my-curl/curl/../libgsasl/pkg/usr/local/lib -lgsasl -L/my-curl/curl/../libidn2/pkg/usr/local/lib -lidn2 -Wl,--pic-executable,-e,mainCRTStartup -Wl,-Map,curl.map' '-DCMAKE_SHARED_LINKER_FLAGS=-static-libgcc -Wl,--nxcompat -Wl,--dynamicbase -Wl,--high-entropy-va -fuse-ld=lld -s -lucrt -Wl,--start-group -lwldap32 -lbcrypt -lbcrypt -lws2_32 -L/my-curl/curl/../libgsasl/pkg/usr/local/lib -lgsasl -L/my-curl/curl/../libidn2/pkg/usr/local/lib -lidn2 -Wl,--image-base,0x150000000 -Wl,--output-def,libcurl-x64.def -Wl,-Map,libcurl-x64.map /my-curl/curl/libcurl.def'
[...]
-- Looking for idn2_lookup_ul in idn2;wldap32;winmm;ws2_32;-lpthread;OpenSSL::SSL;OpenSSL::Crypto;/my-curl/nghttp2/pkg/usr/local/lib/libnghttp2.a;/my-curl/ngtcp2/pkg/usr/local/lib/libngtcp2.a;/my-curl/ngtcp2/pkg/usr/local/lib/libngtcp2_crypto_openssl.a;/my-curl/nghttp3/pkg/usr/local/lib/libnghttp3.a
-- Looking for idn2_lookup_ul in idn2;wldap32;winmm;ws2_32;-lpthread;OpenSSL::SSL;OpenSSL::Crypto;/my-curl/nghttp2/pkg/usr/local/lib/libnghttp2.a;/my-curl/ngtcp2/pkg/usr/local/lib/libngtcp2.a;/my-curl/ngtcp2/pkg/usr/local/lib/libngtcp2_crypto_openssl.a;/my-curl/nghttp3/pkg/usr/local/lib/libnghttp3.a - found
-- Looking for 13 include files /my-curl/libssh2/pkg/usr/local/include/libssh2.h, ..., idn2.h
-- Looking for 13 include files /my-curl/libssh2/pkg/usr/local/include/libssh2.h, ..., idn2.h - found
[...]
+ wine64 pkg/usr/local/bin/curl.exe --version
+ tee curl-x64.txt
curl 7.83.1 (x86_64-w64-mingw32) libcurl/7.83.1 OpenSSL/3.0.4 (Schannel) zlib/1.2.12 brotli/1.0.9 libidn2/2.3.2 libssh2/1.10.0 nghttp2/1.47.0 ngtcp2/0.6.0 nghttp3/0.5.0 libgsasl/1.10.0
Release-Date: 2022-05-11
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI TLS-SRP UnixSockets
```

**Comment by Keitagit-kun** [comment]
I renamed directory to not have spaces: `C:\Temp\curl-7.83.1-OpenSSL3` but still not found
`Looking for idn2_lookup_ul in idn2;wldap32;winmm;ws2_32;C:/Temp/curl-7.83.1-OpenSSL3/deps/lib/libcares.a;OpenSSL::SSL;OpenSSL::Crypto;C:/Temp/curl-7.83.1-OpenSSL3/deps/lib/libnghttp2.a
Looking for idn2_lookup_ul in idn2;wldap32;winmm;ws2_32;C:/Temp/curl-7.83.1-OpenSSL3/deps/lib/libcares.a;OpenSSL::SSL;OpenSSL::Crypto;C:/Temp/curl-7.83.1-OpenSSL3/deps/lib/libnghttp2.a - not found`
maybe my idn-related setup is wrong? Fornow I combined it like this:
`# Check for idn2
option(USE_LIBIDN2 "Use libidn2 for IDN support" ON)
if(USE_LIBIDN2)
  set(CMAKE_C_FLAGS "$${CMAKE_C_FLAGS} -I${prefix}/include")
  set(CMAKE_EXE_LINKER_FLAGS "$${CMAKE_EXE_LINKER_FLAGS} -L${PREFIX}/lib -lidn2")
  set(CMAKE_SHARED_LINKER_FLAGS "$${CMAKE_SHARED_LINKER_FLAGS} -L${PREFIX}/lib -lidn2")
  set(CMAKE_STATIC_LINKER_FLAGS "$${CMAKE_STATIC_LINKER_FLAGS} -L${PREFIX}/lib -lidn2")
  
  check_library_exists_concat("idn2" idn2_lookup_ul HAVE_LIBIDN2)
else()
  set(HAVE_LIBIDN2 OFF)
endif()`

double "$" - is only to counter "Double subscripts: use braces to clarify" error in preview

**Comment by dmilith** [comment]
This is fixed for me in 7.84.0. Thanks guys
